pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/FieldIndexConfigs.java

```diff
@@ -0,0 +1,121 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.segment.spi.index;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.databind.JsonNode;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.apache.pinot.spi.config.table.IndexConfig;
+
+
+/**
+ * FieldIndexConfigs are a map like structure that relates index types with their configuration, providing a type safe
+ * interface.
+ *
+ * This class can be serialized into a JSON object whose keys are the index type ids using Jackson, but cannot be
+ * serialized back. A custom Jackson deserializer could be provided if needed.
+ */
+public class FieldIndexConfigs {
+
+  public static final FieldIndexConfigs EMPTY = new FieldIndexConfigs(new HashMap<>());
+
+  private final Map<IndexType, IndexConfig> _configMap;
+
+  private FieldIndexConfigs(Map<IndexType, IndexConfig> configMap) {
+    _configMap = Collections.unmodifiableMap(configMap);
+  }
+
+  /**
+   * Returns the configuration associated with the given index type, which will be null if there is no configuration for
+   * that index type.
+   */
+  @JsonIgnore
+  public <C extends IndexConfig, I extends IndexType<C, ?, ?>> C getConfig(I indexType) {
+    IndexConfig config = _configMap.get(indexType);
+    if (config == null) {
+      return indexType.getDefaultConfig();
+    }
+    return (C) config;
+  }
+
+  /*
+  This is used by Jackson when this object is serialized. Each entry of the map will be directly contained in the
+  JSON object, with the key name as the key in the JSON object and the result of serializing the key value as the value
+  in the JSON object.
+   */
+  @JsonAnyGetter
+  public Map<String, JsonNode> unwrapIndexes() {
+    Function<Map.Entry<IndexType, IndexConfig>, JsonNode> serializer =
+        entry -> entry.getKey().serialize(entry.getValue());
+    return _configMap.entrySet().stream()
+        .filter(e -> e.getValue() != null)
+        .collect(Collectors.toMap(entry -> entry.getKey().getId(), serializer));
+  }
+
+  public static class Builder {
+    private final Map<IndexType, IndexConfig> _configMap;
+
+    public Builder() {
+      _configMap = new HashMap<>();
+    }
+
+    public Builder(FieldIndexConfigs other) {
+      _configMap = new HashMap<>(other._configMap);
+    }
+
+    public <C extends IndexConfig, I extends IndexType<C, ?, ?>> Builder add(I indexType, @Nullable C config) {
```

Comment:
We can remove it and delegate on `undeclare`

Reply:
Hmm, should we add the default config when it is not explicitly configured?

Reply:
It isn't necessary and it would force us to have larger maps for no reason. It is not necessary because the `getConfig` method is implemented as:

```java
  public <C extends IndexConfig, I extends IndexType<C, ?, ?>> C getConfig(I indexType) {
    IndexConfig config = _configMap.get(indexType);
    if (config == null) {
      return indexType.getDefaultConfig();
    }
    return (C) config;
  }
```

The size of the map wouldn't be an actual memory problem, but:
- It would force us to iterate over all the index types registered whenever we create one of these objects
- We would therefore need to add a dependency to `StandardIndexes`. This pattern were we have to (ab)use of static references is not something I really like, so I'm trying to limit it as much as possible.
- This would make debugging (or logging) quite more difficult, as these objects will have one entry per index instead of one entry per explicitly define index.